### PR TITLE
[SPARK-35556][SQL] Remove close HiveClient's SessionState

### PR DESCRIPTION
**Pull request by LuciferYang**
### What changes were proposed in this pull request?

It will not generate `tmpOutputFile`, `tmpErrOutputFile` and `sessionDirs` since [SPARK-35286](https://issues.apache.org/jira/browse/SPARK-35286). So we can remove `HiveClientImpl.closeState` to avoid these exceptions:
```
java.lang.NoSuchMethodError: org.apache.hadoop.hive.ql.session.SessionState.getTmpErrOutputFile()Ljava/io/File
```

### Why are the changes needed?

1. Avoid incompatible exceptions.
2. Remove useless code.

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?

- Pass the GitHub Action
- Manual test：

Execute

```
mvn clean install -DskipTests -pl sql/hive -am
mvn test -pl sql/hive -DwildcardSuites=org.apache.spark.sql.hive.client.VersionsSuite -Dtest=none
```

**Before**

```
Run completed in 17 minutes, 18 seconds.
Total number of tests run: 867
Suites: completed 2, aborted 0
Tests: succeeded 867, failed 0, canceled 0, ignored 1, pending 0
All tests passed.
15:04:02.407 WARN org.apache.hadoop.hive.metastore.ObjectStore: Version information not found in metastore. hive.metastore.schema.verification is not enabled so recording the schema version 2.3.0
15:04:02.408 WARN org.apache.hadoop.hive.metastore.ObjectStore: setMetaStoreSchemaVersion called but recording version is disabled: version = 2.3.0, comment = Set by MetaStore yangjie01@0.2.30.21
15:04:02.441 WARN org.apache.hadoop.hive.metastore.ObjectStore: Failed to get database default, returning NoSuchObjectException
15:04:03.140 ERROR org.apache.spark.util.Utils: Uncaught exception in thread shutdown-hook-0
java.lang.NoSuchMethodError: org.apache.hadoop.hive.ql.session.SessionState.getTmpErrOutputFile()Ljava/io/File;
	at org.apache.spark.sql.hive.client.HiveClientImpl.$anonfun$closeState$1(HiveClientImpl.scala:168)
	at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:23)
	at org.apache.spark.sql.hive.client.HiveClientImpl.$anonfun$withHiveState$1(HiveClientImpl.scala:312)
	at org.apache.spark.sql.hive.client.HiveClientImpl.liftedTree1$1(HiveClientImpl.scala:243)
	at org.apache.spark.sql.hive.client.HiveClientImpl.retryLocked(HiveClientImpl.scala:242)
	at org.apache.spark.sql.hive.client.HiveClientImpl.withHiveState(HiveClientImpl.scala:292)
	at org.apache.spark.sql.hive.client.HiveClientImpl.closeState(HiveClientImpl.scala:158)
	at org.apache.spark.sql.hive.client.HiveClientImpl.$anonfun$new$1(HiveClientImpl.scala:175)
	at org.apache.spark.util.SparkShutdownHook.run(ShutdownHookManager.scala:214)
	at org.apache.spark.util.SparkShutdownHookManager.$anonfun$runAll$2(ShutdownHookManager.scala:188)
	at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:23)
	at org.apache.spark.util.Utils$.logUncaughtExceptions(Utils.scala:1994)
	at org.apache.spark.util.SparkShutdownHookManager.$anonfun$runAll$1(ShutdownHookManager.scala:188)
	at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:23)
	at scala.util.Try$.apply(Try.scala:213)
	at org.apache.spark.util.SparkShutdownHookManager.runAll(ShutdownHookManager.scala:188)
	at org.apache.spark.util.SparkShutdownHookManager$$anon$2.run(ShutdownHookManager.scala:178)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
15:04:03.141 WARN org.apache.hadoop.util.ShutdownHookManager: ShutdownHook '$anon$2' failed, java.util.concurrent.ExecutionException: java.lang.NoSuchMethodError: org.apache.hadoop.hive.ql.session.SessionState.getTmpErrOutputFile()Ljava/io/File;
java.util.concurrent.ExecutionException: java.lang.NoSuchMethodError: org.apache.hadoop.hive.ql.session.SessionState.getTmpErrOutputFile()Ljava/io/File;
	at java.util.concurrent.FutureTask.report(FutureTask.java:122)
	at java.util.concurrent.FutureTask.get(FutureTask.java:206)
	at org.apache.hadoop.util.ShutdownHookManager.executeShutdown(ShutdownHookManager.java:124)
	at org.apache.hadoop.util.ShutdownHookManager$1.run(ShutdownHookManager.java:95)
Caused by: java.lang.NoSuchMethodError: org.apache.hadoop.hive.ql.session.SessionState.getTmpErrOutputFile()Ljava/io/File;
	at org.apache.spark.sql.hive.client.HiveClientImpl.$anonfun$closeState$1(HiveClientImpl.scala:168)
	at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:23)
	at org.apache.spark.sql.hive.client.HiveClientImpl.$anonfun$withHiveState$1(HiveClientImpl.scala:312)
	at org.apache.spark.sql.hive.client.HiveClientImpl.liftedTree1$1(HiveClientImpl.scala:243)
	at org.apache.spark.sql.hive.client.HiveClientImpl.retryLocked(HiveClientImpl.scala:242)
	at org.apache.spark.sql.hive.client.HiveClientImpl.withHiveState(HiveClientImpl.scala:292)
	at org.apache.spark.sql.hive.client.HiveClientImpl.closeState(HiveClientImpl.scala:158)
	at org.apache.spark.sql.hive.client.HiveClientImpl.$anonfun$new$1(HiveClientImpl.scala:175)
	at org.apache.spark.util.SparkShutdownHook.run(ShutdownHookManager.scala:214)
	at org.apache.spark.util.SparkShutdownHookManager.$anonfun$runAll$2(ShutdownHookManager.scala:188)
	at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:23)
	at org.apache.spark.util.Utils$.logUncaughtExceptions(Utils.scala:1994)
	at org.apache.spark.util.SparkShutdownHookManager.$anonfun$runAll$1(ShutdownHookManager.scala:188)
	at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:23)
	at scala.util.Try$.apply(Try.scala:213)
	at org.apache.spark.util.SparkShutdownHookManager.runAll(ShutdownHookManager.scala:188)
	at org.apache.spark.util.SparkShutdownHookManager$$anon$2.run(ShutdownHookManager.scala:178)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
```

**After**

```
Run completed in 11 minutes, 41 seconds.
Total number of tests run: 867
Suites: completed 2, aborted 0
Tests: succeeded 867, failed 0, canceled 0, ignored 1, pending 0
All tests passed.
```

### CODE REVIEW

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala [hunk 1]
```diff
@@ -165,9 +165,15 @@ private[hive] class HiveClientImpl(
...
       state.setTmpOutputFile(null)
     }
-    if (state.getTmpErrOutputFile != null) {
-      state.getTmpErrOutputFile.delete()
-      state.setTmpErrOutputFile(null)
+    // SPARK-35556: there are some UTs testing multiple versions of hive such as
+    // `org.apache.spark.sql.hive.client.VersionsSuite`, so we need to ensure that
+    // the `getTmpErrOutputFile` method exists in the hive version being tested,
+    // because the `getTmpErrOutputFile` method was added after hive 2.0.
+    if (state.getClass.getMethods.exists(_.getName == "getTmpErrOutputFile")) {
```
Comment (LuciferYang) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala:172]: cc @yaooqinn Do you have any better suggestions

Comment (yaooqinn) [reply sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala:172]: move it to Shim?

Comment (LuciferYang) [reply sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala:172]: thx~ sounds a good idea

**Comment by LuciferYang** [comment]
ping @yaooqinn @dongjoon-hyun @HyukjinKwon

**Comment by HyukjinKwon** [comment]
cc @wangyum

**Review by wangyum — COMMENTED** [review]
It seems that we do not need to deal with these session directories now, because they will not be generated after this pr: https://github.com/apache/spark/pull/32410.
cc @yaooqinn

**Comment by yaooqinn** [comment]
+1 for Yuming's opinion

**Comment by LuciferYang** [comment]
@wangyum  #32410  merged to master 26 days ago, but I still found this Error log in latest jenkins daily test 

image

[https://amplab.cs.berkeley.edu/jenkins/job/spark-master-test-maven-hadoop-3.2/2995/consoleFull](https://amplab.cs.berkeley.edu/jenkins/job/spark-master-test-maven-hadoop-3.2/2995/consoleFull)

**Comment by LuciferYang** [comment]
> It seems that we do not need to deal with these session directories now, because they will not be generated after this pr: #32410.

@wangyum @yaooqinn 

I added a log to check whether `TmpOutputFile` and `TmpErrOutputFile` exist. The code is as follows：


```
if (state.getTmpOutputFile != null) {
      logWarning(s"version = $version, TmpOutputFile ${state.getTmpOutputFile}, " +
        s"exists = ${state.getTmpOutputFile.exists()}")
      state.getTmpOutputFile.delete()
      state.setTmpOutputFile(null)
    }
    if (shim.getTmpErrOutputFile(state) != null) {
      logWarning(s"version = $version, TmpErrOutputFile ${state.getTmpErrOutputFile}, " +
        s"exists = ${state.getTmpErrOutputFile.exists()}")
      state.getTmpErrOutputFile.delete()
      state.setTmpErrOutputFile(null)
    }
```

run `mvn test -pl sql/hive -DwildcardSuites=org.apache.spark.sql.hive.client.VersionsSuite -Dtest=none` and the logs as follows:

```
5:40:16.564 WARN org.apache.spark.sql.hive.client.HiveClientImpl: version = v3_0, TmpOutputFile /spark-source/sql/hive/target/tmp/yangjie01/29465188-6f23-4ff6-b2cd-4106ed6d049d6046166454442575351.pipeout, exists = true
15:40:16.567 WARN org.apache.spark.sql.hive.client.HiveClientImpl: version = v3_0, TmpErrOutputFile /spark-source/sql/hive/target/tmp/yangjie01/29465188-6f23-4ff6-b2cd-4106ed6d049d4242252845976863603.pipeout, exists = true
15:40:16.590 WARN org.apache.spark.sql.hive.client.HiveClientImpl: version = v3_1, TmpOutputFile /spark-source/sql/hive/target/tmp/yangjie01/34d3ac01-ea7c-45a9-888f-94dbfdc39e004387486968654558925.pipeout, exists = true
15:40:16.591 WARN org.apache.spark.sql.hive.client.HiveClientImpl: version = v3_1, TmpErrOutputFile /spark-source/sql/hive/target/tmp/yangjie01/34d3ac01-ea7c-45a9-888f-94dbfdc39e004195215296688492649.pipeout, exists = true
15:40:16.610 WARN org.apache.spark.sql.hive.client.HiveClientImpl: version = v2_3, TmpOutputFile /spark-source/sql/hive/target/tmp/yangjie01/8f609286-87a7-48fb-aaab-f9f903c750a31878401772230169748.pipeout, exists = true
15:40:16.615 WARN org.apache.spark.sql.hive.client.HiveClientImpl: version = v2_3, TmpErrOutputFile /spark-source/sql/hive/target/tmp/yangjie01/8f609286-87a7-48fb-aaab-f9f903c750a38392023963543176532.pipeout, exists = true
15:40:16.642 WARN org.apache.spark.sql.hive.client.HiveClientImpl: version = v3_0, TmpOutputFile /spark-source/sql/hive/target/tmp/yangjie01/d30bdbc6-0c2f-4306-9b1e-3b4aad535e158638013945143377928.pipeout, exists = true
15:40:16.643 WARN org.apache.spark.sql.hive.client.HiveClientImpl: version = v3_0, TmpErrOutputFile /spark-source/sql/hive/target/tmp/yangjie01/d30bdbc6-0c2f-4306-9b1e-3b4aad535e151163344546187356244.pipeout, exists = true
15:40:20.928 WARN org.apache.hadoop.hive.metastore.ObjectStore: Version information not found in metastore. hive.metastore.schema.verification is not enabled so recording the schema version 2.3.0
15:40:20.929 WARN org.apache.hadoop.hive.metastore.ObjectStore: setMetaStoreSchemaVersion called but recording version is disabled: version = 2.3.0, comment = Set by MetaStore yangjie01@0.2.30.21
15:40:20.961 WARN org.apache.hadoop.hive.metastore.ObjectStore: Failed to get database default, returning NoSuchObjectException
15:40:21.511 WARN org.apache.spark.sql.hive.client.HiveClientImpl: version = v2_2, TmpOutputFile /spark-source/sql/hive/target/tmp/yangjie01/b412fffa-94f7-4e4e-9737-76ebbcea68657537376491865215575.pipeout, exists = true
15:40:21.514 WARN org.apache.spark.sql.hive.client.HiveClientImpl: version = v2_2, TmpErrOutputFile /spark-source/sql/hive/target/tmp/yangjie01/b412fffa-94f7-4e4e-9737-76ebbcea68652630341764089541067.pipeout, exists = true
15:40:21.554 WARN org.apache.spark.sql.hive.client.HiveClientImpl: version = v2_3, TmpOutputFile /spark-source/sql/hive/target/tmp/yangjie01/78b164aa-4a19-4290-92f3-382754f6b5eb8817210522107381307.pipeout, exists = true
15:40:21.556 WARN org.apache.spark.sql.hive.client.HiveClientImpl: version = v2_3, TmpErrOutputFile /spark-source/sql/hive/target/tmp/yangjie01/78b164aa-4a19-4290-92f3-382754f6b5eb9018760134240849099.pipeout, exists = true
15:40:21.588 WARN org.apache.spark.sql.hive.client.HiveClientImpl: version = v12, TmpOutputFile /spark-source/sql/hive/target/tmp/yangjie01/f0f0993d-e112-4d72-96bc-72f1bd7b3dce8389333243580648925.pipeout, exists = true
15:40:21.598 WARN org.apache.spark.sql.hive.client.HiveClientImpl: version = v1_0, TmpOutputFile /spark-source/sql/hive/target/tmp/yangjie01/b5028767-1dcb-460e-a73a-f1040a03a4087784566790767080481.pipeout, exists = true
15:40:21.622 WARN org.apache.spark.sql.hive.client.HiveClientImpl: version = v2_2, TmpOutputFile /spark-source/sql/hive/target/tmp/yangjie01/27eb39e0-33a6-4841-997f-f1fe739674332330608174582929049.pipeout, exists = true
15:40:21.623 WARN org.apache.spark.sql.hive.client.HiveClientImpl: version = v2_2, TmpErrOutputFile /spark-source/sql/hive/target/tmp/yangjie01/27eb39e0-33a6-4841-997f-f1fe739674338623579564031605430.pipeout, exists = true
15:40:21.673 WARN org.apache.spark.sql.hive.client.HiveClientImpl: version = v2_0, TmpOutputFile /spark-source/sql/hive/target/tmp/yangjie01/5a622559-dd15-4847-a64c-c1f463bf10181812901792092274832.pipeout, exists = false
15:40:21.675 WARN org.apache.spark.sql.hive.client.HiveClientImpl: version = v2_0, TmpErrOutputFile /spark-source/sql/hive/target/tmp/yangjie01/5a622559-dd15-4847-a64c-c1f463bf10182583426440911855811.pipeout, exists = false
15:40:21.705 WARN org.apache.spark.sql.hive.client.HiveClientImpl: version = v2_1, TmpOutputFile /spark-source/sql/hive/target/tmp/yangjie01/1694b243-75dd-4a1a-b6cb-1ccc0df4275c2848140385223458989.pipeout, exists = false
15:40:21.707 WARN org.apache.spark.sql.hive.client.HiveClientImpl: version = v2_1, TmpErrOutputFile /spark-source/sql/hive/target/tmp/yangjie01/1694b243-75dd-4a1a-b6cb-1ccc0df4275c7935159698768455602.pipeout, exists = false
15:40:21.728 WARN org.apache.spark.sql.hive.client.HiveClientImpl: version = v1_2, TmpOutputFile /spark-source/sql/hive/target/tmp/yangjie01/5afe0523-06c8-4a43-b50f-cb67460c570c5947222341246452210.pipeout, exists = true
15:40:21.754 WARN org.apache.spark.sql.hive.client.HiveClientImpl: version = v2_0, TmpOutputFile /spark-source/sql/hive/target/tmp/yangjie01/e30b548b-d41d-4a33-b120-533babe886622691756388180862769.pipeout, exists = false
15:40:21.755 WARN org.apache.spark.sql.hive.client.HiveClientImpl: version = v2_0, TmpErrOutputFile /spark-source/sql/hive/target/tmp/yangjie01/e30b548b-d41d-4a33-b120-533babe886627545865308020288420.pipeout, exists = false
15:40:21.759 WARN org.apache.spark.sql.hive.client.HiveClientImpl: version = v2_1, TmpOutputFile /spark-source/sql/hive/target/tmp/yangjie01/6dff4ebd-9801-42a6-a286-22d98852a5545896208236220536311.pipeout, exists = false
15:40:21.760 WARN org.apache.spark.sql.hive.client.HiveClientImpl: version = v2_1, TmpErrOutputFile /spark-source/sql/hive/target/tmp/yangjie01/6dff4ebd-9801-42a6-a286-22d98852a5543427804008723015422.pipeout, exists = false
15:40:21.771 WARN org.apache.spark.sql.hive.client.HiveClientImpl: version = v1_1, TmpOutputFile /spark-source/sql/hive/target/tmp/yangjie01/a73dc77c-0b2b-4985-9954-7582fdf170b27143760359965874400.pipeout, exists = true
15:40:21.802 WARN org.apache.spark.sql.hive.client.HiveClientImpl: version = v1_2, TmpOutputFile /spark-source/sql/hive/target/tmp/yangjie01/2190ef5b-430c-422c-81d7-da47946895ba7983579376143486857.pipeout, exists = true
15:40:21.813 WARN org.apache.spark.sql.hive.client.HiveClientImpl: version = v1_0, TmpOutputFile /spark-source/sql/hive/target/tmp/yangjie01/d72db119-ed5a-4820-8928-adab2c7e701e8329597961731395160.pipeout, exists = true
15:40:21.815 WARN org.apache.spark.sql.hive.client.HiveClientImpl: version = v1_1, TmpOutputFile /spark-source/sql/hive/target/tmp/yangjie01/4332fcf3-b786-4aa7-9f00-fd0f168445593159479301496984340.pipeout, exists = true
15:40:21.837 WARN org.apache.spark.sql.hive.client.HiveClientImpl: version = v14, TmpOutputFile /spark-source/sql/hive/target/tmp/yangjie01/e961b814-183c-4f65-a5e6-8acb7823a1733564943126481452919.pipeout, exists = true
15:40:21.867 WARN org.apache.spark.sql.hive.client.HiveClientImpl: version = v13, TmpOutputFile /spark-source/sql/hive/target/tmp/yangjie01/6b7223d3-cef6-4d27-b25a-07af0cd5ad868143994489599912504.pipeout, exists = true
15:40:21.869 WARN org.apache.spark.sql.hive.client.HiveClientImpl: version = v14, TmpOutputFile /spark-source/sql/hive/target/tmp/yangjie01/2ca0568c-89c7-4f54-a347-f0192ede7cc35477674386745689474.pipeout, exists = true
15:40:21.886 WARN org.apache.spark.sql.hive.client.HiveClientImpl: version = v12, TmpOutputFile /spark-source/sql/hive/target/tmp/yangjie01/1f4a0b25-b701-4d9a-93c7-105b4e293a565965652575218891295.pipeout, exists = true
15:40:21.887 WARN org.apache.spark.sql.hive.client.HiveClientImpl: version = v13, TmpOutputFile /spark-source/sql/hive/target/tmp/yangjie01/d311cc01-6a43-4f76-853d-30107015212b8320766062103240048.pipeout, exists = true
15:40:21.894 WARN org.apache.spark.sql.hive.client.HiveClientImpl: version = v12, TmpOutputFile /spark-source/sql/hive/target/tmp/yangjie01/5f58ec76-86fb-486f-980e-b0bf9ad03f372514319116057676135.pipeout, exists = true
```

It seems that `TmpOutputFile` and `TmpErrOutputFile` were created。

**Comment by wangyum** [comment]
It seems `TmpOutputFile` and `TmpErrOutputFile`  are generated by [SessionState.start(state)](https://github.com/apache/spark/blob/ebb4858f7185c6525adc4b23bc89f0a8262bf940/sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala#L829).

I think we can remove these lines:
https://github.com/apache/spark/blob/ebb4858f7185c6525adc4b23bc89f0a8262bf940/sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala#L158-L175
and add `state.close()` to `runHive`.

**Comment by wangyum** [comment]
@LuciferYang Please update the PR title and PR description.

**Comment by LuciferYang** [comment]
@wangyum Thanks for your guidance.

**Review by wangyum — APPROVED** [review]

**Review by yaooqinn — APPROVED** [review]
LGTM

**Comment by LuciferYang** [comment]
thx @wangyum @yaooqinn @HyukjinKwon

**Comment by wangyum** [comment]
Merged to master.

**Pull request closed without merging**